tests/testlib/s2n_ktls_test_utils.c

```diff
@@ -0,0 +1,239 @@
+        MOCK_IO_ENSURE_REF(buf);
+        size_t len = msg->msg_iov[count].iov_len;
+
+        /* If we fail to write to stuffer then return blocked */
+        if (s2n_stuffer_write_bytes(&io_ctx->data_buffer, buf, len) < 0) {
+            errno = EAGAIN;
+            return -1;
+        }
+
+        total_len += len;
+    }
+    if (total_len) {
+        /* write record_type and len after data was written successfully. ancillary_buffer is
+         * growable so this operation should always succeed. */
+        MOCK_IO_GUARD(s2n_stuffer_write_uint8(&io_ctx->ancillary_buffer, record_type));
+        MOCK_IO_GUARD(s2n_stuffer_write_uint16(&io_ctx->ancillary_buffer, total_len));
+    }
+
+    return total_len;
+}
+
+/* In userspace TLS, s2n first reads the header to determine the length of next record
+ * and then reads the entire record into conn->in. In kTLS its not possible to know
+ * the length of the next record. Instead the socket returns the minimum of
+ * bytes-requested and data-available; reading multiple consecutive records if they
+ * are of the same type. */
+ssize_t s2n_test_ktls_recvmsg_stuffer_io(struct s2n_connection *conn, struct msghdr *msg, uint8_t *record_type)
+{
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(conn->recv_io_context);
+    POSIX_ENSURE_REF(msg);
+    POSIX_ENSURE_REF(msg->msg_iov);
+
+    struct s2n_test_ktls_io_stuffer *io_ctx = (struct s2n_test_ktls_io_stuffer *) conn->recv_io_context;
+    POSIX_ENSURE_REF(io_ctx);
+    /* Assert ancillary_buffer is growable, which simplifies IO logic. Blocking IO can
+     * be mocked by restricting the data_buffer. */
+    POSIX_ENSURE(io_ctx->ancillary_buffer.growable, S2N_ERR_IO);
+    io_ctx->invoked_count++;
+
+    /* s2n only receives using msg_iovlen of 1 */
+    MOCK_IO_ENSURE(msg->msg_iovlen == 1);
+
+    uint8_t *buf = msg->msg_iov->iov_base;
+    MOCK_IO_ENSURE_REF(buf);
+
+    /* There is no data available so return blocked */
+    if (!s2n_stuffer_data_available(&io_ctx->ancillary_buffer)) {
+        errno = EAGAIN;
+        return -1;
+    }
+
+    ssize_t total_read = 0;
+    /* updated as partial or multiple records are read */
+    size_t updated_requested_len = msg->msg_iov->iov_len;
+    /* track two record_types since its possible to read multiple records of the same type */
+    *record_type = 0;
+    uint8_t next_record_type = 0;
+    while (*record_type == next_record_type) {
+        /* read record_type and number of bytes available in the next record */
+        MOCK_IO_GUARD(s2n_stuffer_read_uint8(&io_ctx->ancillary_buffer, record_type));
+        uint16_t n_avail = 0;
+        MOCK_IO_GUARD(s2n_stuffer_read_uint16(&io_ctx->ancillary_buffer, &n_avail));
+        POSIX_ENSURE_LTE(n_avail, s2n_stuffer_data_available(&io_ctx->data_buffer));
+
+        size_t n_read = MIN(updated_requested_len, n_avail);
+        MOCK_IO_ENSURE(n_read > 0);
+
+        /* we have already verified that there is more data, so this should succeed */
```

Comment:
Why verify it early if this method will check it anyway?

Reply:
Again prob a comment we can just remove. We check early so that we can return the EAGAIN errno. I do this outside of the loop so easier to understand and we really only need to check it the first time when we havnt read any data.